src/Control/SessionHandler/FileSessionHandler.php

```diff
@@ -0,0 +1,351 @@
+     * @inheritDoc
+     * Called instead of write if session.lazy_write is enabled and no data has changed for this session.
+     */
+    public function updateTimestamp(#[SensitiveParameter] string $id, string $data): bool
+    {
+        $path = $this->getFilePath($id);
+        $filesystem = new Filesystem();
+
+        try {
+            // The $data isn't actually meant to be used here and seems to only be in the interface method signature
+            // because when wrapping native handlers with the SessionHandler class you can't call parent::updateTimestamp()
+            // and must therefore call parent::write($id, $data) - the data is needed there to avoid clearing out the
+            // session data.
+            $time = DBDatetime::now()->getTimestamp();
+            $filesystem->touch($path, $time, $time);
+            return true;
+        } catch (IOException $e) {
+            Injector::inst()->get(LoggerInterface::class)->error(
+                'Could not update timestamp for session file: ' . $this->getSafeExceptionMessage($e, $id)
+            );
+            return false;
+        }
+    }
+
+    /**
+     * Set the path sessions will be saved in, along with optional additional configuration.
+     * @throws InvalidArgumentException if $path is invalid (e.g. has too many semicolons)
+     */
+    private function setSavePath(string $path): void
+    {
+        // Handle configuration for depth and mode arguments
+        // See https://www.php.net/manual/en/session.configuration.php#ini.session.save-path for details
+        if (str_contains($path, ';')) {
+            $parts = explode(';', $path);
+            $numParts = count($parts);
+            if ($numParts > 3) {
+                throw new InvalidArgumentException('$path was invalid');
+            }
+
+            // Set n depth
+            $this->numSubDirs = (int) array_shift($parts);
+
+            // Mode is the 2nd item in the config. Usually will be like '600' but we need true octal e.g. 0600
+            if ($numParts === 3) {
+                $this->mode = $this->ensureValueIsOctal(array_shift($parts));
+            }
+
+            // The actual path is the last part of the configuration.
+            $path = array_shift($parts);
+        }
+        $this->baseDir = $path;
+    }
+
+    /**
+     * Get the absolute path to the file which this session's data will be stored in.
+     */
+    private function getFilePath(#[SensitiveParameter] string $id): string
+    {
+        $subParts = str_split(substr($id, 0, $this->numSubDirs));
+        return Path::join(...[$this->baseDir, ...$subParts, FileSessionHandler::SESSION_FILE_PREFIX . $id]);
+    }
+
+    /**
+     * Take a numeric string from save_path configuration and make sure it's a PHP octal integer.
+     */
+    private function ensureValueIsOctal(string $value): int
+    {
+        if (strlen($value) < 4) {
+            $value = '0' . $value;
+        }
+        return intval($value, 8);
```

Comment:
```suggestion
        return octdec($value);
```

Reply:
Nice! Didn't know about that function. Done.